### ISSUE TITLE
diy lesson plan for gradient-descent-style fit for Model 4

### ISSUE THREAD

**Issue by bbartling** [bbartling/hvac-optimal-start-math-playground#1]
We want to find **τ** and **k** that minimize the **mean squared error** between `t_pred` and `t_actual`.

---

## What “iterative optimizer” + “convergence control” mean (in this context)

* **Iterative optimizer**
  You *guess* τ and k, compute error, nudge them, and repeat:

  1. Start with τ₀, k₀
  2. Compute predictions `t_pred(τ,k)`
  3. Compute error vs. `t_actual`
  4. Use the error to update τ,k slightly
  5. Loop until “good enough”

* **Convergence control**
  Rules that decide:

  * how big each step is (`learning_rate`)
  * when to stop (loss stops improving, or max iterations)
  * what to do if things blow up (e.g., NaNs, huge loss)

---

## Easy Python: gradient-descent-style fit for Model 4

This is **plain NumPy**, no SciPy required. You can literally paste this into a `model4_fit_demo.py` as a teaching script.

```python
import numpy as np

# ---------------------------------------------------------
# 1. Fake "history" data (ΔT, t_actual)
#    Replace with real data later.
# ---------------------------------------------------------
delta_t = np.array([2.0, 4.0, 6.0, 8.0, 10.0], dtype=float)
t_actual = np.array([10.0, 18.0, 27.0, 35.0, 42.0], dtype=float)

# ---------------------------------------------------------
# 2. Model 4 form: t_pred = tau * ln(1 + k * ΔT)
#    We will fit tau and k using gradient descent.
# ---------------------------------------------------------

# Initial guesses
tau = 20.0   # minutes
k   = 0.05   # 1/°F

# Learning rates (step sizes)
lr_tau = 1e-2
lr_k   = 1e-4

max_iters = 5000
tol = 1e-8
prev_loss = None

def safe_log_term(k, dt):
    """
    Avoids invalid log arguments (<= 0).
    Clamps 1 + k*ΔT to a small positive value if it goes too low.
    """
    x = 1.0 + k * dt
    x = np.maximum(x, 1e-6)  # clamp for safety
    return x

for it in range(max_iters):
    # ----- forward pass -----
    log_arg = safe_log_term(k, delta_t)
    t_pred = tau * np.log(log_arg)

    # Residuals and loss (mean squared error)
    residuals = t_pred - t_actual
    loss = np.mean(residuals ** 2)

    # Convergence control: stop if loss not changing much
    if prev_loss is not None and abs(prev_loss - loss) < tol:
        print(f"Converged at iter {it}, loss={loss:.6f}")
        break
    prev_loss = loss

    # ----- backward pass: gradients -----
    # d(t_pred)/d(tau) = ln(1 + k * ΔT)
    d_t_pred_d_tau = np.log(log_arg)

    # d(t_pred)/d(k) = tau * (ΔT / (1 + k * ΔT))
    d_t_pred_d_k = tau * (delta_t / log_arg)

    # dLoss/dθ = 2 * mean(residual * d(t_pred)/dθ)
    dL_d_tau = 2.0 * np.mean(residuals * d_t_pred_d_tau)
    dL_d_k   = 2.0 * np.mean(residuals * d_t_pred_d_k)

    # ----- parameter update (gradient descent step) -----
    tau -= lr_tau * dL_d_tau
    k   -= lr_k   * dL_d_k

    # Optional: simple debug print every 500 iters
    if it % 500 == 0:
        print(
            f"iter={it:4d}  loss={loss:.4f}  "
            f"tau={tau:.4f}  k={k:.5f}"
        )

print("\nFinal fit:")
print(f"  tau ≈ {tau:.4f} minutes")
print(f"  k   ≈ {k:.5f} 1/°F")
print(f"  final loss = {loss:.6f}")

# Quick check: predict for each ΔT with fitted params
log_arg = safe_log_term(k, delta_t)
t_pred_final = tau * np.log(log_arg)
print("\nΔT    t_actual   t_predicted")
for dt, ta, tp in zip(delta_t, t_actual, t_pred_final):
    print(f"{dt:4.1f}  {ta:8.2f}   {tp:10.2f}")
```

### Where “iterative optimizer” + “convergence control” show up in this code

* **Iterative optimizer**

  * The `for it in range(max_iters):` loop
  * Each pass:

    * compute predictions
    * compute loss
    * compute gradients
    * update `tau` and `k`

* **Convergence control**

  * `max_iters` = hard cap on how long we’ll iterate
  * `tol` + `abs(prev_loss - loss) < tol` = “stop if not improving”
  * `safe_log_term` = prevents the optimizer from stepping into invalid log() territory
  * Small `lr_tau`, `lr_k` = prevent steps that are too big and would blow the loss up
